subsys/tracing/tracing_backend_adsp_memory_window.c

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2024 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ctype.h>
+#include <zephyr/kernel.h>
+#include <string.h>
+#include <tracing_core.h>
+#include <tracing_buffer.h>
+#include <tracing_backend.h>
+
+#include <adsp_memory.h>
+#include <adsp_debug_window.h>
+
+/* structure of memory window */
+struct mem_window {
+	uint32_t head_offset;	/* offset of the first not used byte in data[] */
+	uint8_t data[];		/* tracing data */
+} __packed __aligned(8);
+
+#define ADSP_TRACING_WINDOW_DATA_SIZE (ADSP_DW_SLOT_SIZE - offsetof(struct mem_window, data))
+
+static volatile struct mem_window *mem_window;
+
+static void tracing_backend_adsp_memory_window_output(
+		const struct tracing_backend *backend,
+		uint8_t *data, uint32_t length)
+{
+	/* copy data to ring buffer,
+	 * to make FW part fast, there's no sync with the data reader
+	 * the reader MUST read data before they got overwritten
+	 */
+	size_t to_copy = ADSP_TRACING_WINDOW_DATA_SIZE - mem_window->head_offset;
+
+	if (to_copy > length)
+		to_copy = length;
```

Comment:
`to_copy = MIN(length, ADSP_TRACING_WINDOW_DATA_SIZE - mem_window->head_offset);` is a cleaner way to express this